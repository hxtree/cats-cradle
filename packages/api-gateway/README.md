# @org-package/api-gateway

![TypeScript](https://shields.io/badge/TypeScript-3178C6?logo=TypeScript&logoColor=FFF&style=flat-square)
![Lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen)

This service brings up bare bone API Gateway Rest APIs, Stages, and Deployments.
Other services then use these to resources via Platform Constructs to trigger
new deployments. This allows multiple stacks to deploy resources to the same API
Gateway. It's kind of similar to a Kong API Gateway AWS Lambda but straight CDK.

- Less costly sandbox/developer environments than using AWS ACM Certificate
  approach, which is ~$500/month.
- Avoids interservice DNS TTL caching issues.
- Easier coordination between multiple microservices as they can share the same
  host.
- Simplifies maintenance of different (mobile, phone, web, etc) backend for
  frontend.

## Stacks

- ApiGateway

## Documentation

- [Architecting Multiple Microservices Behind Single Domain](https://aws.amazon.com/blogs/compute/architecting-multiple-microservices-behind-a-single-domain-with-amazon-api-gateway/)
- [Crossstack References](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/walkthrough-crossstackref.html)
