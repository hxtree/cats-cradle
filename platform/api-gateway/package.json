{
  "name": "@cats-cradle/api-gateway",
  "module": "commonjs",
  "version": "0.1.2",
  "bin": {
    "app": "bin/app.js"
  },
  "scripts": {
    "dev": "nest start --debug --watch",
    "artifact": "create-artifact @cats-cradle/api-gateway",
    "build": "tsc",
    "build:all": "rush build -t .",
    "watch": "tsc -w",
    "test": "jest --runInBand",
    "test:cov": "jest --coverage --runInBand",
    "lint": "eslint --format visualstudio \"./{stacks,src}/*.{ts,tsx}\"  --fix",
    "lint:ci": "eslint --format visualstudio \"./{stacks,src}/*.{ts,tsx}\"  --fix-dry-run",
    "cdk:bootstrap": "cdk bootstrap",
    "cdk:diff": "cdk diff",
    "cdk:synth": "cdk synth",
    "cdk:hotswap": "cdk deploy --hotswap",
    "cdk:deploy": "cdk deploy",
    "cdk:watch": "cdk watch",
    "cdk:destroy": "cdk destroy",
    "depcheck": "npx depcheck",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "aws-lambda": "1.0.7",
    "class-transformer": "0.5.1",
    "class-validator": "0.13.2",
    "@nestjs/common": "9.4.0",
    "@nestjs/core": "9.2.1",
    "@nestjs/terminus": "9.1.1",
    "@nestjs/platform-express": "9.3.9",
    "aws-serverless-express": "~3.4.0",
    "@vendia/serverless-express": "4.10.4",
    "express": "4.18.2",
    "node-cache": "5.1.2",
    "reflect-metadata": "0.1.13",
    "rxjs": "7.8.1",
    "mongoose": "6.9.1",
    "source-map-support": "~0.5.21",
    "@nestjs/swagger": "~7.1.13",
    "uuid": "~9.0.1",
    "@aws-cdk/core": "~1.182.0",
    "constructs": "^10.2.70",
    "aws-cdk-lib": "2.95.0",
    "construct": "~1.0.0"
  },
  "devDependencies": {
    "@cats-cradle/create-artifact": "2.0.0",
    "@cats-cradle/constructs": "workspace:*",
    "@cats-cradle/base-nodejs": "workspace:*",
    "@types/jest": "29.5.5",
    "@types/node": "20.8.2",
    "@types/aws-lambda": "~8.10.123",
    "jest": "29.7.0",
    "ts-jest": "29.1.1",
    "typescript": "5.1.3",
    "ts-node": "10.9.1",
    "esbuild": "~0.17.5",
    "fs": "~0.0.1-security",
    "path": "~0.12.7",
    "@types/express": "~4.17.18",
    "@types/uuid": "~9.0.6",
    "eslint": "8.44.0",
    "@cats-cradle/eslint-config": "1.0.6"
  }
}
