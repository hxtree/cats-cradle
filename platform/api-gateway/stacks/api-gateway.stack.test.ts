import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ApiGatewayStack } from './api-gateway.stack';

describe('ApiGatewayStack', () => {
  it('should match snapshot test', () => {
    const app = new cdk.App();
    const stack = new ApiGatewayStack(app, 'MyTestStack');
    const template = Template.fromStack(stack);

    // TODO consider reworking to use DNS instead of incrementing and deploying each time
    const customSnapshotSerializer = {
      test: (val: any) => typeof val === 'string',
      print: (val: any) => val.replace(/deployment\d{8}T\d{9}Z[^:]+/, 'deployment_UNIQUE_HASH'),
    };
    expect.addSnapshotSerializer(customSnapshotSerializer);

    expect(template).toMatchSnapshot();
  });
});
