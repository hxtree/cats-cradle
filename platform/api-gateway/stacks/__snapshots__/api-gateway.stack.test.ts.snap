// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiGatewayStack should match snapshot test 1`] = `
{
  Parameters: {
    BootstrapVersion: {
      Default: /cdk-bootstrap/hnb659fds/version,
      Description: Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip],
      Type: AWS::SSM::Parameter::Value<String>,
    },
  },
  Resources: {
    MyTestStackdeployment_UNIQUE_HASH: {
      DependsOn: [
        MyTestStackwebapigatewaymockANY084D1838,
      ],
      Properties: {
        RestApiId: {
          Ref: MyTestStackwebapigateway11D00362,
        },
      },
      Type: AWS::ApiGateway::Deployment,
    },
    MyTestStackstageA41D5650: {
      DependsOn: [
        MyTestStackwebapigatewayAccountF9D6F084,
      ],
      Properties: {
        DeploymentId: {
          Ref: MyTestStackdeployment_UNIQUE_HASH,
        },
        RestApiId: {
          Ref: MyTestStackwebapigateway11D00362,
        },
        StageName: default,
      },
      Type: AWS::ApiGateway::Stage,
    },
    MyTestStackwebapigateway11D00362: {
      Properties: {
        Name: web-api-gateway,
      },
      Type: AWS::ApiGateway::RestApi,
    },
    MyTestStackwebapigatewayAccountF9D6F084: {
      DeletionPolicy: Retain,
      DependsOn: [
        MyTestStackwebapigateway11D00362,
      ],
      Properties: {
        CloudWatchRoleArn: {
          Fn::GetAtt: [
            MyTestStackwebapigatewayCloudWatchRole3A2B89EF,
            Arn,
          ],
        },
      },
      Type: AWS::ApiGateway::Account,
      UpdateReplacePolicy: Retain,
    },
    MyTestStackwebapigatewayCloudWatchRole3A2B89EF: {
      DeletionPolicy: Retain,
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Action: sts:AssumeRole,
              Effect: Allow,
              Principal: {
                Service: apigateway.amazonaws.com,
              },
            },
          ],
          Version: 2012-10-17,
        },
        ManagedPolicyArns: [
          {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs,
              ],
            ],
          },
        ],
      },
      Type: AWS::IAM::Role,
      UpdateReplacePolicy: Retain,
    },
    MyTestStackwebapigatewayid2F65D424: {
      Properties: {
        Description: Web API Gateway Rest API ID,
        Name: web-api-gateway-rest-api-id,
        Type: String,
        Value: {
          Ref: MyTestStackwebapigateway11D00362,
        },
      },
      Type: AWS::SSM::Parameter,
    },
    MyTestStackwebapigatewaymock96E57734: {
      Properties: {
        ParentId: {
          Fn::GetAtt: [
            MyTestStackwebapigateway11D00362,
            RootResourceId,
          ],
        },
        PathPart: mock,
        RestApiId: {
          Ref: MyTestStackwebapigateway11D00362,
        },
      },
      Type: AWS::ApiGateway::Resource,
    },
    MyTestStackwebapigatewaymockANY084D1838: {
      Properties: {
        AuthorizationType: NONE,
        HttpMethod: ANY,
        Integration: {
          IntegrationResponses: [
            {
              StatusCode: 200,
            },
          ],
          PassthroughBehavior: NEVER,
          RequestTemplates: {
            application/json: { "statusCode": 200 },
          },
          Type: MOCK,
        },
        MethodResponses: [
          {
            StatusCode: 200,
          },
        ],
        ResourceId: {
          Ref: MyTestStackwebapigatewaymock96E57734,
        },
        RestApiId: {
          Ref: MyTestStackwebapigateway11D00362,
        },
      },
      Type: AWS::ApiGateway::Method,
    },
    MyTestStackwebapigatewayresourceid87DDADFE: {
      Properties: {
        Description: Web Gateway Resource ID,
        Name: web-api-gateway-root-resource-id,
        Type: String,
        Value: {
          Fn::GetAtt: [
            MyTestStackwebapigateway11D00362,
            RootResourceId,
          ],
        },
      },
      Type: AWS::SSM::Parameter,
    },
    MyTestStackwebapigatewayv1resourceFF514C41: {
      Properties: {
        ParentId: {
          Fn::GetAtt: [
            MyTestStackwebapigateway11D00362,
            RootResourceId,
          ],
        },
        PathPart: v1,
        RestApiId: {
          Ref: MyTestStackwebapigateway11D00362,
        },
      },
      Type: AWS::ApiGateway::Resource,
    },
    MyTestStackwebapigatewayv1resourceid8E5650DB: {
      Properties: {
        Description: Web API Gateway V1 Resource ID,
        Name: web-api-gateway-v1-resource-id,
        Type: String,
        Value: {
          Ref: MyTestStackwebapigatewayv1resourceFF514C41,
        },
      },
      Type: AWS::SSM::Parameter,
    },
    MyTestStackwebmockresourceid87D417E8: {
      Properties: {
        Description: Web Mock Resource ID,
        Name: web-mockapi-gateway-root-resource-id,
        Type: String,
        Value: {
          Fn::GetAtt: [
            MyTestStackwebapigateway11D00362,
            RootResourceId,
          ],
        },
      },
      Type: AWS::SSM::Parameter,
    },
  },
  Rules: {
    CheckBootstrapVersion: {
      Assertions: [
        {
          Assert: {
            Fn::Not: [
              {
                Fn::Contains: [
                  [
                    1,
                    2,
                    3,
                    4,
                    5,
                  ],
                  {
                    Ref: BootstrapVersion,
                  },
                ],
              },
            ],
          },
          AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.,
        },
      ],
    },
  },
}
`;
